fix(proxy-rewrite): query string discarded when use_real_request_uri_…

YapWC opened the pull request:
…unsafe is true

When use_real_request_uri_unsafe is set to true, url that was requested with query string, arguments is not parsed. Expected to parse the arguments just like same default behavior when use_real_request_uri_unsafe is set to false.

The real issue lies in the fact that when use_real_request_uri_unsafe is set to true and conf.uri is provided, the core.utils.resolve_var output back an upstream uri that does not include query string. Therefore we need to concatenate the query string for the case if conf.uri is true.

### Description

<!-- Please include a summary of the change and which issue is fixed. -->
<!-- Please also include relevant motivation and context. -->

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
-->
Fixes #12623 

### Checklist

- [x] I have explained the need for this PR and the problem it solves
- [x] I have explained the changes or the new features added to this PR
- [x] I have added tests corresponding to this change
- [ ] I have updated the documentation to reflect this change
- [x] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

<!--

Note

1. Mark the PR as draft until it's ready to be reviewed.
2. Always add/update tests for any changes unless you have a good reason.
3. Always update the documentation to reflect the changes made in the PR.
4. Make a new commit to resolve conversations instead of `push -f`.
5. To resolve merge conflicts, merge master instead of rebasing.
6. Use "request review" to notify the reviewer after making changes.
7. Only a reviewer can mark a conversation as resolved.

-->

YapWC committed:
fix(proxy-rewrite): query string discarded when use_real_request_uri_unsafe is true

When use_real_request_uri_unsafe is set to true, url that was requested with query string, arguments is not parsed. Expected to parse the arguments just like same default behavior when use_real_request_uri_unsafe is set to false.

The reason this problem exist is because `if ctx.var.is_args == "?" then` statement is inside `if not conf.use_real_request_uri_unsafe then` which then bypasses args parsing.

Baoyuantop reviewed (commented):
If the plugin only configures "use_real_request_uri_unsafe": true, it will cause the query to be concatenated repeatedly.

YapWC commented:
Hi @Baoyuantop alright I think I understand what you mean. I think it is better for me to debug it line by line in this case. For your side normally how do you run debugging/printing code output?

YapWC committed:
Merge branch 'apache:master' into fix/proxy-rewrite-discarded-query-string

YapWC committed:
fix: query string bypass when use unsafe true

The real issue lies in the fact that when use_real_request_uri_unsafe is set to true and conf.uri is provided, the core.utils.resolve_var output back an upstream uri that does not include query string. Therefore we need to concatenate the query string for the case if conf.uri is true.

The previous commit does not solve the issue as the issue is not related in the `if not conf.use_real_request_uri_unsafe then` statement.

YapWC committed:
Merge branch 'apache:master' into fix/proxy-rewrite-discarded-query-string

YapWC committed:
fix: Lint for line too long

moonming reviewed (changes requested):
Hi @YapWC, thank you for fixing the query string issue in proxy-rewrite!

Query strings being discarded when using `use_real_request_uri` is a real bug that affects production use. 

**Please**:
1. Rebase on the latest master
2. Add edge case tests: empty query strings, query strings with special characters, multiple `?` in the URI
3. Verify the fix works correctly with both `use_real_request_uri_unsafe` variants

Looking forward to the update!

Baoyuantop commented:
Hi @YapWC, following up on the previous review comments. Please let us know if you have any updates. Thank you.

nic-6443 commented on t/plugin/proxy-rewrite.t:
```diff
@@ -1231,3 +1231,95 @@ GET /hello
+
+
+
+=== TEST 46: rewrite uri args
```
The earlier revision of this PR double-appended the query when only `use_real_request_uri_unsafe: true` was set (no `uri`) — that is fixed now since `query_string` is only appended in the `conf.uri` branch, but none of the tests actually pin that scenario down. TEST 8/9 in proxy-rewrite3.t cover unsafe mode without a query string, and the new tests here all set `uri` together with the flag. Since this exact case already regressed once during review, it would be good to add a test with just `use_real_request_uri_unsafe: true` (no `uri`/`regex_uri`) and a request like `/hello?q=apisix`, asserting the query reaches the upstream exactly once.

AlinsRan committed:
fix(proxy-rewrite): merge query string with '&' and simplify comments

Address review feedback: when conf.uri already carries its own query string,
merge the original query with '&' instead of concatenating a second '?'.
Simplify the added comments.

AlinsRan committed:
Merge remote-tracking branch 'origin/master' into sync/pr-12843

AlinsRan committed:
test(proxy-rewrite): cover conf.uri own query, special chars, multiple '?' and both variants

Address review feedback on the query-string fix:
- conf.uri carrying its own query merged with '&' (no double '?')
- use_real_request_uri_unsafe only (no conf.uri): query reaches upstream once
- query with special characters and multiple '?' passed through unchanged
- same conf.uri with use_real_request_uri_unsafe disabled, verifying both
  variants keep the query string consistently

AlinsRan committed:
test(proxy-rewrite): reindex proxy-rewrite3.t test numbers

Fix non-sequential TEST numbers (45/46 -> 42/43) so the reindex lint check passes.

AlinsRan committed:
Merge remote-tracking branch 'origin/master' into sync/pr-12843
